src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,492 @@
+        var constantMappers = new List<ConstantsMapper>();
+        var nodesRemapping = new Dictionary<SyntaxNode, SyntaxNode>();
+        var variableValues = contextInformation.VariableValues;
+        var builderToAnalysisContext = contextInformation.BuilderToAnalysisContext;
+        var builderNodesProcessed = contextInformation.BuilderNodesProcessed;
+        var variableValuesList = contextInformation.VariableValuesList;
+
+        foreach (var builderNode in builderNodeToVariableValue.Keys)
+        {
+            var variableValue = builderNodeToVariableValue[builderNode];
+            var analysisContext = variableValue.ExpressionAnalysisContext;
+            var rewrittenBuildersNode = analysisContext.Node.RewrittenExpression;
+            var constantsMapper = analysisContext.Node.ConstantsRemapper;
+            argumentTypeName = analysisContext.Node.ArgumentTypeName;
+            nodesRemapping.Add(builderNode, rewrittenBuildersNode);
+            constantMappers.Add(constantsMapper);
+        }
+
+        var rewrittenSyntaxNode = syntaxNode.ReplaceNodes(nodesRemapping.Keys, (n, _) => nodesRemapping[n]);
+
+        if (isCompound)
+        {
+            var rewrittenAssignmentExpression = rewrittenSyntaxNode as AssignmentExpressionSyntax;
+            var binaryExpressionSyntaxKind = GenerateBinaryExpressionSyntaxKind(syntaxKind);
+            rewrittenSyntaxNode = SyntaxFactory.BinaryExpression(binaryExpressionSyntaxKind, rewrittenAssignmentExpression.Left, rewrittenAssignmentExpression.Right);
+        }
+
+        var rewrittenBuildersExpression = rewrittenSyntaxNode switch
+        {
+            AssignmentExpressionSyntax @assignmentExpression => @assignmentExpression.Right,
+            VariableDeclaratorSyntax @variableDeclarator => @variableDeclarator.Initializer.Value,
+            _ => rewrittenSyntaxNode
+        };
+
+        var (finalBuildersExpression, finalConstantMapper) = ConstantsMapperComposer.Compose(rewrittenBuildersExpression, constantMappers);
+        return new ExpressionAnalysisContext(new ExpressionAnalysisNode(rewrittenBuildersExpression, argumentTypeName, finalBuildersExpression, finalConstantMapper, new Location[] { }));
+    }
+
+    private static bool ProcessDeclaration(VariableDeclaratorSyntax variableDeclarator, ContextInformation contextInformation, HashSet<string> dirtyVariables, bool displayDiagnostics)
+    {
+        var variableValues = contextInformation.VariableValues;
+        var variableValuesList = contextInformation.VariableValuesList;
+        var builderNodesProcessed = contextInformation.BuilderNodesProcessed;
+
+        var variableName = variableDeclarator.Identifier.ValueText;
+        var evaluationResult = PreEvaluate(variableDeclarator, contextInformation, null, displayDiagnostics, out var variableValue);
+        var canEvaluate = evaluationResult.CanEvaluate;
+        var syntaxNodeToVariableValueMap = evaluationResult.SyntaxNodeToVariableValueMap;
```

Comment:
No need to extract most of the variables, like:
canEvaluate,
syntaxNodeToVariableValueMap...